packages/dds/tree/src/test/openPolymorphism.spec.ts

```diff
@@ -0,0 +1,273 @@
+		 *
+		 * TODO: document (and enforce/detect) when how late it is safe to modify array's used as allowed types.
+		 * These docs should ideally align with how late lazy type lambdas are evaluated (when the tree configuration is constructed, or an instance is made, which ever is first? Maybe define schema finalization?)
+		 */
+		const ItemTypes: ItemSchema[] = [];
+
+		// -------------
+		// Library using an Item
+
+		class Container extends sf.array("Container", ItemTypes) {}
+
+		// -------------
+		// Library defining an item
+
+		ItemTypes.push(TextItem);
+
+		// -------------
+		// Example use of container with generic code and down casting
+
+		const container = new Container();
+
+		// If we don't do anything special, the insertable type is never, so a cast is required to insert content.
+		container.insertAtStart(new TextItem({ text: "", location: { x: 0, y: 0 } }) as never);
+
+		// Items read from the container are typed as Item and have thew expected APIs:
+		const first = container[0];
+		first.foo();
+		first.location.x += 1;
+
+		// Down casting works as normal.
+		if (Tree.is(first, TextItem)) {
+			assert.equal(first.text, "foo");
+		}
+	});
+
+	it("mutable static registry, error cases", () => {
+		const ItemTypes: ItemSchema[] = [];
+		class Container extends sf.array("Container", ItemTypes) {}
+
+		// Not added to registry
+		// ItemTypes.push(TextItem);
+
+		const container = new Container();
+
+		// Should error due to out of schema content
+		assert.throws(
+			() =>
+				container.insertAtStart(new TextItem({ text: "", location: { x: 0, y: 0 } }) as never),
+			validateUsageError(/schema/),
+		);
+
+		// Modifying registration too late should error
+		assert.throws(() => ItemTypes.push(TextItem));
+	});
+
+	it("mutable static registry, recursive case", () => {
+		const ItemTypes: ItemSchema[] = [];
+
+		// Example recursive item implementation
+		class Container extends sf.array("Container", ItemTypes) {}
+		class ContainerItem extends sf.object("ContainerItem", {
+			...itemFields,
+			container: Container,
+		}) {
+			public static readonly description = "Text";
+			public static default(): TextItem {
+				return new TextItem({ text: "", location: { x: 0, y: 0 } });
+			}
+
+			public foo(): void {}
+		}
+
+		ItemTypes.push(ContainerItem);
+
+		const container = new Container();
+
+		container.insertAtStart(
+			new ContainerItem({ container: [], location: { x: 0, y: 0 } }) as never,
+		);
+	});
+
+	it("mutable static registry, safer editing API", () => {
+		const ItemTypes: ItemSchema[] = [];
+		class Container extends sf.object("Container", {
+			child: sf.optional(customizeSchemaTyping(ItemTypes).simplified<Item>()),
+		}) {}
+
+		ItemTypes.push(TextItem);
+
+		const container = new Container({ child: undefined });
+		const container2 = new Container({ child: TextItem.default() });
+
+		// Enabled by customizeSchemaTyping
+		container.child = TextItem.default();
+		container.child = undefined;
+	});
+
```

Comment:
I would love to try something like this:
```
interface A {
  type: "a";
  foo: string;
}
interface B {
  type: "b";
  bar: number;
}
interface C {
  easy: number
}
type Combo = A | B;
interface Root {
  faz: Combo;
  c: C;
}

class SyncedRoot extends sf.object(...) implements Root {}
...

const a: A = { ... }
const root: Root = {...}
const syncedRoot = new SyncedRoot(root);
syncedRoot.faz = new SyncedA(a);
```

Reply:
Cases like that where Combo is explicitly referring the types in the union work fine, even without this change. The problem that is being addressed and demoed in this file is "open "polymorphism where you need to somehow express "Combo" without referencing the actual members in it leaving the implicit union "open" for extension (in such a setup the implementation refer to what they are implementing, in this case "Combo", like if Combo and SyncCombo were interfaces things could implement, not a union of types).